src/app/modules/startup/internal/state_wrapper.nim

```diff
@@ -0,0 +1,56 @@
+import NimQml
+import state
+
+QtObject:
+  type StateWrapper* = ref object of QObject
+    stateObj: State
+
+  proc delete*(self: StateWrapper) =
+    self.QObject.delete
+
+  proc newStateWrapper*(): StateWrapper =
+    new(result, delete)
+    result.QObject.setup()
+
+  proc setStateObj*(self: StateWrapper, stateObj: State) =
+    self.stateObj = stateObj
+
+  proc getStateObj*(self: StateWrapper): State =
+    return self.stateObj
+
+  proc getFlowType(self: StateWrapper): string {.slot.} =
+    if(self.stateObj.isNil):
+      return $FlowType.General
+    return $self.stateObj.flowType()
+  QtProperty[string] flowType:
+    read = getFlowType
+
+  proc getStateType(self: StateWrapper): string {.slot.} =
+    if(self.stateObj.isNil):
+      return $StateType.NoState
+    return $self.stateObj.stateType()
+  QtProperty[string] stateType:
+    read = getStateType
+
+  proc getDisplayBackButton(self: StateWrapper): bool {.slot.} =
+    if(self.stateObj.isNil):
+      return false
+    return self.stateObj.displayBackButton()
+  QtProperty[bool] displayBackButton:
```

Comment:
Got it, if it works then it works :D

But this seems to be wrong: https://github.com/status-im/status-desktop/blob/0a62ff8144593f2c803dd65fc925262f25d9d744/src/app/modules/startup/view.nim#L64-L72

`currentStartupState` will always point to the same QVariant with the same object behind. Is it guaranteed that qml will reevaluate bindings if it detects nothing really changed (property value stays the same even though notify signal was emitted)?

What makes more sense is:
```
 proc setCurrentStartupState*(self: View, state: State) = 
   self.currentStartupState = newStateObj(state) 
   self.currentStartupStateChanged() 
```
only then properties inside `StateWrapper` can be truly constant (after removing `setStateObj`)

Reply:
@osmaczko whenever we set new state via `self.currentStartupState.setStateObj(state)` and emit `self.currentStartupStateChanged()` all props of `StateWrapper` which are exposed to qml as `currentStartupState` will be reevaluated for sure. `currentStartupStateVariant` is present only because of Nim limitations, in c++ will be all the same, without `currentStartupStateVariant` instance, and props of `StateWrapper` will be `CONSTANT ` `Q_PROPERTY`.

Reply:
Still not my point. Translating it to cpp, current solution:

```cpp
StateWrapper {
    Q_OBJECT
    Q_PROPERTY(QString foo READ getFoo CONSTANT)
public:
    StateWrapper() = default;
    void setStateObject(std::unique_ptr<StateObject> stateObject) { mStateObject = stateObject; }
    QString getFoo() { return mStateObject->foo(); }
private:
    std::unique_ptr<StateObject> mStateObject;
}
```
In this case, foo should **NOT** be a CONSTANT property, it would be constant if it never changes after StateWrapper construction, but here it can change and will change when `setStateObject` is invoked. With this implementation `StateWrapper` can be easily misused, resulting in `foo` property being different than one would expect.

```cpp
View {
    Q_PROPERTY(StateWrapper* state READ getState NOTIFY stateChanged)
public:
   StateWrapper* getState() { return mStateWrapper.get(); }
   void setCurrentStartupState(std::unique_ptr<StateObject> stateObject) {
       mStateWrapper.setStateObject(stateObject);
       emit stateChanged();
   }
private:
    std::unique_ptr<StateWrapper> mStateWrapper;
}
```
This is also wrong, because `StateWrapper*` would point to the exactly same object in memory as before, so the property does not change at all. In other words: `getState` will return the same value every time `stateChanged()` is triggered. The correct way would be (assuming that `setStateObject` is be removed from `StateWrapper`):
```
   void setCurrentStartupState(std::unique_ptr<StateObject> stateObject) {
       mStateWrapper = std::make_unique<StateWrapper>(stateObject);
       emit stateChanged();
   }
```

Reply:
@osmaczko it's up to one who calls it, but I agree that we should minimize the risk. Also there are multiple ways of doing this. 
This is one of possible implementations in c++, where each `Q_PROPERTY` of `StateWrapper` is `CONSTANT`:
```
void setCurrentStartupState(std::unique_ptr<StateObject> stateObject) {
       mStateWrapper = std::make_unique<StateWrapper>(stateObject);
       emit stateChanged();
   }
```

Another option is where `currentState` is constant, but props of `StateWrapper` are not, but share the same signal, that way we're safe that anybody who calls `setCurrentStartupState` won't care about emitting signal:
```
Q_PROPERTY(StateWrapper* currentState READ getCurrentState CONSTANT)

void setCurrentState(std::unique_ptr<StateObject> stateObject) {
       mStateWrapper.setStateObject(std::move(stateObject));
}
```

in `StateWrapper` have this function:
```
Q_PROPERTY(QString foo1 READ getFoo1 NOTIFY stateWrapperChanged)
Q_PROPERTY(QString foo2 READ getFoo2 NOTIFY stateWrapperChanged)
Q_PROPERTY(QString foo3 READ getFoo3 NOTIFY stateWrapperChanged)

void setStateObject(std::unique_ptr<StateObject> stateObject) { 
      mStateObject = std::move(stateObject); 
      emit stateWrapperChanged();
}
``` 

But what we were doing in Nim, in all places around, is exactly what is written there. So `currentStartupStateVariant` is initialized once and it points to `currentStartupState` which is also initialized once, and we're just emitting `currentStartupStateChanged` when we set new underlying object for `currentStartupState` and pay attention to emit `currentStartupStateChanged` signal in that case. I don't see any issues if exposed `currentStartupStateVariant` points always to the same `currentStartupState` which points to the different underlying objects if we emit signal to notify all parties which are reading properties that underlying object has been changed. But since your concern is that somebody may call directly `setStateObj`, I can change it to the option 2 from above or even to the option 1.

Reply:
> I don't see any issues if exposed currentStartupStateVariant points always to the same currentStartupState which points to the different underlying objects if we emit signal to notify all parties which are reading properties that underlying object has been changed.

My concern here was "Is it guaranteed that qml will reevaluate bindings if it detects nothing really changed (property value stays the same even though notify signal was emitted)?" But since you confirmed qml reevaluates bindings in that case, there is no issue then.

> But since your concern is that somebody may call directly setStateObj, I can change it to the option 2 from above or even to the option 1.

Exactly, this was my point :P Option 2 looks better because it makes `StateWrapper` hard to be used wrongly. Also, in that case, `self.currentStartupStateChanged()` will not be needed anymore (I believe :thinking:)

Reply:
@osmaczko updated to option 2 to be perfectly safe. ;)